phoenix-core/src/main/java/org/apache/phoenix/coprocessor/UngroupedAggregateRegionObserver.java

```diff
@@ -1189,40 +1237,168 @@ private Delete generateDeleteMarkers(List<Cell> row) {
+                    indexMaintainer.getEmptyKeyValueQualifier(), ts, VERIFIED_BYTES);
+            int cellCount = 0;
+            for (List<Cell> cells : indexPut.getFamilyCellMap().values()) {
+                if (cells == null) {
+                    break;
+                }
+                for (Cell expectedCell : cells) {
+                    Cell actualCell = indexRow.getColumnLatestCell(CellUtil.cloneFamily(expectedCell),
+                            CellUtil.cloneQualifier(expectedCell));
+                    if (actualCell == null) {
+                        exceptionMessage = "Index verify failed - Missing cell " + indexHTable.getName();
+                        throw new IOException(exceptionMessage);
+                    }
+                    // Check all columns
+                    if (!CellUtil.matchingValue(actualCell, expectedCell)) {
+                        exceptionMessage = "Index verify failed - Not matching cell value - " + indexHTable.getName();
+                        throw new IOException(exceptionMessage);
+                    }
+                    cellCount++;
+                }
+            }
+            if (cellCount != indexRow.rawCells().length) {
+                exceptionMessage = "Index verify failed - Expected to find " + cellCount + " cells but got "
+                        + indexRow.rawCells().length + " cells - " + indexHTable.getName();
+                throw new IOException(exceptionMessage);
+            }
+        }
+
+        private void verifyIndexRows(ArrayList<KeyRange> keys) throws IOException {
+            int expectedRowCount = keys.size();
+            ScanRanges scanRanges = ScanRanges.createPointLookup(keys);
+            Scan indexScan = new Scan();
+            indexScan.setTimeRange(scan.getTimeRange().getMin(), scan.getTimeRange().getMax());
+            scanRanges.initializeScan(indexScan);
+            SkipScanFilter skipScanFilter = scanRanges.getSkipScanFilter();
+            indexScan.setFilter(skipScanFilter);
+            int rowCount = 0;
+            try (ResultScanner resultScanner = indexHTable.getScanner(indexScan)) {
+                for (Result result = resultScanner.next(); (result != null); result = resultScanner.next()) {
+                    Put dataPut = dataPutMap.get(result.getRow());
+                    if (dataPut == null) {
+                        exceptionMessage = "Index verify failed - Missing data row - " + indexHTable.getName();
+                        throw new IOException(exceptionMessage);
+                    }
+                    verifySingleIndexRow(result, dataPut);
+                    rowCount++;
+                }
+            } catch (Throwable t) {
+                ServerUtil.throwIOException(indexHTable.getName().toString(), t);
+            }
+            if (rowCount != expectedRowCount) {
+                exceptionMessage = "Index verify failed - Missing index rows Expected : " + expectedRowCount +
```

Comment:
I was wondering if we can have a test for this expected message

Reply:
I will do that